Automatic forward to configured Forwarded Endpoint in case of no matching mocked response and no default response

mariog-tw opened the pull request:
This feature is related to the following open issue: #288 

For both SOAP and Rest projects we implemented the following scenario:

Given a configured method/operation with any response strategy and a Forwarded Endpoint and no default response
When a request does not match any mocked response
Then the request will be forwarded to the configured Forwarded Endpoint

In order to be retrocompatible with the existing behaviour, the forwarding takes place only if no mocked response is matched and no default response is configured.

karldahlgren commented:
Thank you so much for your contribution. I'm a little bit split on what I think about the suggested changes. I'm afraid that it might lead to some unwanted behaviour. There is currently nothing that indicates for user that a non-match will result in a request being forwarded. I'm wondering if there should be some sort of option that would make this more clear for the user that this would occur. For example, there could be a checkbox setting in the method that basically says to forward request that it is unable to match a mocked response for. What do you think?

karldahlgren commented on web/web-mock/web-mock-rest/src/main/java/com/castlemock/web/mock/rest/controller/mock/RestClient.java:
```diff
@@ -0,0 +1,70 @@
+package com.castlemock.web.mock.rest.controller.mock;
```
Missing license header

karldahlgren commented on web/web-mock/web-mock-rest/src/main/java/com/castlemock/web/mock/rest/controller/mock/AbstractRestServiceController.java:
```diff
@@ -262,45 +242,11 @@ private RestResponse forwardRequest(final RestRequest request,
-                    .build();
-        } catch (IOException exception) {
-            LOGGER.error("Unable to forward request", exception);
+        Optional<RestResponse> optionalRestResponse = restClient.getResponse(request, restMethod);
```
Could make this a little bit shorter by doing something like this>
`return restClient.getResponse(request, restMethod).orElseThrow(() -> new RestException("Unable to forward request to configured endpoint");)`

karldahlgren commented on web/web-mock/web-mock-rest/src/main/java/com/castlemock/web/mock/rest/controller/mock/AbstractRestServiceController.java:
```diff
@@ -26,24 +25,14 @@
 import com.castlemock.model.core.utility.parser.ExternalInputBuilder;
 import com.castlemock.model.core.utility.parser.TextParser;
 import com.castlemock.model.core.utility.parser.expression.argument.ExpressionArgument;
-import com.castlemock.model.mock.rest.domain.RestEvent;
```
I personally prefer to specify what is being imported instead of using *. It is more clear what is being imported.

karldahlgren commented on web/web-mock/web-mock-rest/src/main/java/com/castlemock/web/mock/rest/controller/mock/AbstractRestServiceController.java:
```diff
@@ -512,6 +458,10 @@ protected RestResponse mockResponse(final RestRequest restRequest,
             }
         }
 
+        if (mockResponse == null && restMethod.getForwardedEndpoint() != null) {
```
Could be nice to check if the restMethod is configured to forward upon no matching mocked responses.

karldahlgren commented on web/web-mock/web-mock-soap/src/main/java/com/castlemock/web/mock/soap/controller/mock/SoapClient.java:
```diff
@@ -0,0 +1,57 @@
+package com.castlemock.web.mock.soap.controller.mock;
```
Missing license header

karldahlgren commented on web/web-mock/web-mock-soap/src/test/java/com/castlemock/web/mock/soap/web/soap/controller/SoapServiceControllerTest.java:
```diff
@@ -19,22 +19,15 @@
-import com.castlemock.model.mock.soap.domain.SoapResourceType;
-import com.castlemock.model.mock.soap.domain.SoapResponseStrategy;
-import com.castlemock.model.mock.soap.domain.SoapXPathExpression;
+import com.castlemock.model.mock.soap.domain.*;
```
Importing with *

karldahlgren reviewed (commented):
Once again, great contribution. Overall it looks good, but I think it would be great if a RestMethod could be configured to use this specific functionality.

mariog-tw commented:
Hi @karldahlgren I pushed some commits addressing the following gaps that were there in the previous ones:

1.  Added checkbox to make automatic forwarding configurable on both Rest Methods and SOAP Operations
2. Changed models and repos classes to accomodate the persistence of this new flag
3. Fixed the imports to be explicit instead of *
4. Added license headers

Hope you can give it a look and let us know in case there are some things we missed and need more attention.
Thanks!

mariog-tw commented on web/web-frontend/src/components/project/rest/method/modal/UpdateMethodModal.js:
```diff
@@ -84,7 +85,8 @@ class UpdateMethodModal extends PureComponent {
                 ...this.state.updateMethod,
-                forwardedEndpoint: forwardedEndpoint
+                forwardedEndpoint: forwardedEndpoint,
+                automaticForward: forwardedEndpoint ? this.state.updateMethod.automaticForward : false
```
This allows us to ensure that if the forwarded endpoint is changed to an empty value then we also disable the automatic forward feature since it wouldn't make sense to forward.

mariog-tw commented on web/web-frontend/src/components/project/rest/method/modal/UpdateMethodModal.js:
```diff
@@ -129,9 +140,11 @@ class UpdateMethodModal extends PureComponent {
                 ...this.state.updateMethod,
-                defaultMockResponseId: defaultMockResponseId
+                defaultMockResponseId: defaultMockResponseId,
+                automaticForward: defaultMockResponseId == "-- select an option --" ? this.state.updateMethod.automaticForward : false
```
Again here we make sure that if the user selects a default mock response then we automatically disable the automatic forward since we would always fallback to this default response if nothing is matching the request.
